java-frontend/src/main/java/org/sonar/java/cfg/CFG.java

```diff
@@ -766,41 +795,26 @@ private void buildTryStatement(TryStatementTree tryStatementTree) {
     }
     Block finallyOrEndBlock = currentBlock;
     Block beforeFinally = createBlock(currentBlock);
-    List<Block> catches = new ArrayList<>();
+    Map<Type, Block> catches = new HashMap<>();
+    TryStatement tryStatement = new TryStatement();
+    tryStatement.successorBlock = finallyOrEndBlock;
+    enclosingTry.push(tryStatement);
     for (CatchTree catchTree : tryStatementTree.catches()) {
       currentBlock = createBlock(finallyOrEndBlock);
       build(catchTree.block());
-      catches.add(currentBlock);
+      catches.put(catchTree.parameter().type().symbolType(), currentBlock);
     }
+    catches.forEach(tryStatement::addCatch);
     currentBlock = beforeFinally;
     build(tryStatementTree.block());
-    linkToCatchBlocks(beforeFinally, catches);
     build((List<? extends Tree>) tryStatementTree.resources());
+    enclosingTry.pop();
     currentBlock = createBlock(currentBlock);
     currentBlock.elements.add(tryStatementTree);
-    if (finallyBlockTree != null) {
-      exitBlocks.pop();
-      if (catches.isEmpty()) {
-        currentBlock.addExitSuccessor(finallyOrEndBlock);
-      }
-    }
-    for (Block catchBlock : catches) {
-      currentBlock.addSuccessor(catchBlock);
-    }
-  }
-
-  protected void linkToCatchBlocks(Block block, List<Block> catches) {
-    Block blockForSuccessor = block;
-    if (block.exitBlock != null && block.exitBlock.isFinallyBlock) {
-      blockForSuccessor = block.exitBlock;
-    }
-    for (Block catchBlock : catches) {
-      blockForSuccessor.addSuccessor(catchBlock);
-    }
   }
 
   private void buildThrowStatement(ThrowStatementTree throwStatementTree) {
-    // FIXME this won't work if it is intended to be caught by a try statement.
+    // TODO check enclosing try to determine jump target
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1135)